src/proof/supernova.rs

```diff
@@ -228,10 +234,65 @@ where
             info!("prove_step {i}");
 
             recursive_snark
-                .prove_step(&pp.pp, step, &step.secondary_circuit())
+                .prove_step(&pp.pp, step, &secondary_circuit)
                 .unwrap();
 
             recursive_snark_option = Some(recursive_snark);
+        };
+
+        if lurk_config(None, None)
+            .perf
+            .parallelism
+            .recursive_steps
+            .is_parallel()
+        {
+            let cc = steps
+                .into_iter()
+                .map(|mf| (mf.program_counter() == 0, Mutex::new(mf)))
+                .collect::<Vec<_>>();
+
+            crossbeam::thread::scope(|s| {
+                s.spawn(|_| {
+                    // Skip the very first circuit's witness, so `prove_step` can begin immediately.
+                    // That circuit's witness will not be cached and will just be computed on-demand.
+
+                    // There are many MultiFrames with PC = 0, each with several inner frames and heavy internal
+                    // paralellism for witness generation. So we do it like on Nova's pipeline.
```

Comment:
Did you measure and observe that this special-casing helps? If it doesn't or is insignificant, I'd argue that the complexity might not justify the handling.

Reply:
We don't have a robust/stressing enough benchmark for NIVC to observe it yet, but I am pretty confident that both options below are worse than what this PR implements:

* Cache all witnesses in parallel
* Cache all witnesses sequentially

#941 demonstrates the first bullet point

To understand the second bullet point, we need to visualize some Multiframe for, say, sha256, in the middle of a sequence of multiframes for Lurk's step function. This multiframe only has one frame (so it can't benefit from multi-frame parallelism) and it encodes some big handwritten circuit that can't benefit from inner parallelism like the Lurk step function does.

Of course, we should measure and confirm it eventually (hence the TODO)